src/cargo/core/gc.rs

```diff
@@ -0,0 +1,550 @@
+                    "gc.auto.max-git-co-age",
+                    auto_config.max_git_co_age.as_deref().unwrap_or("1 month"),
+                )?;
+                self.max_git_db_age = newer_time_span_for_config(
+                    self.max_git_db_age,
+                    "gc.auto.max-git-db-age",
+                    auto_config.max_git_db_age.as_deref().unwrap_or("3 months"),
+                )?;
+            }
+            if matches!(kind, AutoGcKind::Target | AutoGcKind::SharedTarget) {
+                bail!("target is unimplemented");
+            }
+        }
+        if let Some(max_download_age) = max_download_age {
+            self.max_src_age = Some(maybe_newer_span(max_download_age, self.max_src_age));
+            self.max_crate_age = Some(maybe_newer_span(max_download_age, self.max_crate_age));
+            self.max_index_age = Some(maybe_newer_span(max_download_age, self.max_index_age));
+            self.max_git_co_age = Some(maybe_newer_span(max_download_age, self.max_git_co_age));
+            self.max_git_db_age = Some(maybe_newer_span(max_download_age, self.max_git_db_age));
+        }
+        Ok(())
+    }
+}
+
+/// The kind of automatic garbage collection to perform.
+///
+/// "Automatic" is the kind of gc performed automatically by Cargo in any
+/// command that is already doing a bunch of work. See [`auto_gc`] for more.
+#[derive(Clone, Debug)]
+pub enum AutoGcKind {
+    /// Automatically clean up the downloaded files *and* the target directory.
+    ///
+    /// This is the mode used by default.
+    All,
+    /// Automatically clean only downloaded files.
+    ///
+    /// This corresponds to `cargo clean --gc=download`.
+    Download,
+    /// Automatically clean only the target directory.
+    ///
+    /// THIS IS NOT IMPLEMENTED.
+    ///
+    /// This corresponds to `cargo clean --gc=target`.
+    Target,
+    /// Automatically clean only the shared target directory.
+    ///
+    /// THIS IS NOT IMPLEMENTED.
+    ///
+    /// This corresponds to `cargo clean --gc=shared-target`.
+    SharedTarget,
+}
+
+impl<'a, 'config> Gc<'a, 'config> {
+    pub fn new(
+        config: &'config Config,
+        global_cache_tracker: &'a mut GlobalCacheTracker,
+    ) -> CargoResult<Gc<'a, 'config>> {
+        let lock = config.acquire_package_cache_lock(CacheLockMode::MutateExclusive)?;
+        Ok(Gc {
+            config,
+            global_cache_tracker,
+            lock,
+        })
+    }
+
+    /// Performs automatic garbage cleaning.
+    ///
+    /// This returns immediately without doing work if garbage collection has
+    /// been performed recently (since `gc.auto.frequency`).
+    fn auto(&mut self, clean_ctx: &mut CleanContext<'config>) -> CargoResult<()> {
+        if !self.config.cli_unstable().gc {
+            return Ok(());
+        }
+        let auto_config = self
+            .config
+            .get::<Option<AutoConfig>>("gc.auto")?
+            .unwrap_or_default();
```

Comment:
Double check but I'm assuming I'm over-optimizing by thinking that for `gc.auto` we should only load as much of the config as is necessary for determining whether GC needs to run or no before loading everything else.

Reply:
I think the overhead of the `.get()` code is fairly heavy in comparison to the actual creation of the value here. If was separate, it would have to go through that mechanism twice in the common case, which I think would be more expensive (or at least a wash).

Reply:
Fair enough

Reply:
Thought of another angle of this: layering.  I assume whatever you get gets layered together, so if we get `gc.auto` that means if you override one field in a config, you mask all of the other fields.  Is that correct?

How much do we care though?

Reply:
If by masking you mean they get set to `None` in the deserialized form, yea that's how tables get merged.

Reply:
Seems sub-optimal to merge on the table level, rather than the field level, but in working on #12889, I realized we already have precedence for this.